src/lib/Libifl/pbs_loadconf.c

```diff
@@ -1109,11 +1112,14 @@ __pbs_loadconf(int reload)
 
 	pbs_conf.loaded = 1;
 
-	if (parse_psi(psi_value) == -1) {
-		fprintf(stderr, "Couldn't find a valid server instance to connect to\n");
-		free(psi_value);
-		goto err;
-	}
+	if (!reload) {
```

Comment:
psi_str was used to get the raw PSI value. If PSI is NULL, just make it NULL and do not have to use a framed value using pbs_default().

Now if you have that, just compare the strings and return when they are the same within parse_psi.

Instead what you have done is to choose not to re-read the values even when reload=TRUE. which is not the case with any other conf parameters and goes against the definition of the function.

Reply:
What is the difference?
pbs_loadconf() does error check, and we can enforce a type of check there. In this case the check is that, we will not reload the value of pbs_conf.psi. I do not see any issue with that. As a precedence, look at pbs_conf.pbs_leaf_routers which is a value that is made-up by loadconf (not exactly what was loaded).

Reply:
pbs_loadconf isn't just meant to parse pbs.conf, it's meant to do it smartly. We create pbs_conf.psi even when PBS_SERVER_INSTANCES is NULL so that the code which needs it doesn't need special case logic for when pbs_conf.psi is NULL, there can be just a single code flow.

About reload: TBH, 'reload' sounds like it was meant to 'reload' a select few pbs.conf parameters, but we didn't implement it that way. That's why some daemons don't even seem to call pbs_loadconf on a HUP (e.g - server's HUP handler just closes and reopens the logs, same with sched). So, even if this is the first parameter that's not reloaded on 'reload', I think it's not wrong to do so. We should reload only those params which make sense to be reloaded, others need a restart.

Reply:
I was referring to psi_str, not pbs_conf.psi. psi_str is used to point to the PSI data and it does not need a framed data. But that does not require to have a special case logic. Now using that, you can just exit if you have the same old data without parsing any further.

You mention how specific daemons behave with the re-read data. Some may act upon it and some may not. That is exactly what I am saying. There is a division of responsibility. pbs_loadconf just reads and makes the data available for the daemons to act upon.

If the daemons do not reload the conf file upon HUP, then isn't it evident that the daemons are responsible to decide that and what is the benefit of adding such a check here?

Now if you want to head in the other direction where only selected parameters are re-read, what will be those parameters? Does it vary from daemon to daemon? So why not leave pbs_loadconf as today which will re-read all parameters upon invocation and leave the rest for the caller.